src/pass.rs

```diff
@@ -0,0 +1,229 @@
+/// Interfacing with WASM.
+
+use std::path::Path;
+use std::fmt;
+use std::iter;
+use std::collections::HashMap;
+use parity_wasm::elements::*;
+
+use util::Either;
+
+#[derive(Debug)]
+pub struct WasmModule {
+    module: Module,
+    function_names: HashMap<usize, String>,
+}
+
+impl WasmModule {
+    pub fn from_file<P: AsRef<Path>>(path: P) -> Result<Self, Error> {
+        let module = deserialize_file(path)?;
+        let mut result = WasmModule::new(module);
+        let function_names = result.exported_function_names();
+        result.function_names = function_names;
+        Ok(result)
+    }
+
+    fn new(module: Module) -> Self {
+        WasmModule {
+            module,
+            function_names: HashMap::new(),
+        }
+    }
+
+    pub fn functions(&self) -> impl Iterator<Item = WasmFunction> {
+        let function_count = self.module.functions_space();
+        if function_count == 0 {
+            return Either::A(iter::empty::<WasmFunction>());
```

Comment:
What is the `A` here?

Reply:
Is it a shorthand for `Left`?

Reply:
Nevermind. I see the self-implelementation of an `Either` in `util.rs`. Can we call it something else? I assumed this was the `Either` in the `futures` crate.

Reply:
It's semantically equivalent to the `Either` in `futures`, I think -- just a more generic `Result`. I can rename `A` and `B` to `Left` and `Right`, though.